### ISSUE TITLE
Move away from xmltoman for manpages?

### ISSUE THREAD

**Issue by rossburton** [lathiat/avahi#527]
xmltoman is basically unmaintained and uses a bespoke XML format.

As an alternative I suggest either sphinx or docutils: both have restructuredtext input and can generate manpages and more, so the pieces of documentation could be consolidated into a single unit which contains the design overview, manpages, etc.

**Comment by evverx** [comment]
> xmltoman is basically unmaintained and uses a bespoke XML format.

While I agree it isn't ideal I'm not sure why it would be an issue at this point. It kind of works, distros don't plan to throw it away as far as I know. The bespoke format probably makes it harder to change the documentation but it is not that there are a lot of contributors who are particularly interested in that anyway.

Realistically it isn't a priority and I don't think `xmltoman` is going to be replaced with anything else here in the foreseeable future (unless patches are sent probably).

**Comment by rossburton** [comment]
Well, the background is that OpenEmbedded hasn't got a recipe for xmltoman yet (because we've used tarballs of avahi, and nothing else has needed it) and I don't want to add a new recipe to core for a tool that is abandoned.

So I might send a patch :)

**Comment by evverx** [comment]
I get why OpenEmbedded wouldn't want to introduce build dependencies that haven't changed for a while into their build pipelines. I'm not sure the other distros would be happy about that though. They all would have to change their packages to point to a new thing.

Personally I'm fine with switching to something else but I'd wait for the others to chime in so that it would be possible to reach a consensus.

**Comment by mbiebl** [comment]
+1 from the Debian side for moving away from xmltoman

**Comment by pemensik** [comment]
From what I have seen, [BIND9](https://gitlab.isc.org/isc-projects/bind9/) could serve as a nice example of converting from docboox to rst files for manual pages. Could be taken as an inspiration. xml files seems too unconvenient to use for documentation generation today. But it seems to me we have priority in cleaning up the code for now.

**Comment by evverx** [comment]
@pemensik if you are in Europe I think it should be possible to apply for https://www.sovereigntechfund.de/programs/bug-resilience. I think the "direct contribution" part (https://www.sovereigntechfund.de/programs/bug-resilience#direct-contributions) can help here to switch to Sphinx. The program focuses on improving maintainability and security (including I assume supply chain security where abandoned dependencies are frowned upon) and getting rid of xmltoman can help with that. I think they can also help with tests, compiler warnings and stuff like that.

I'm 99% sure avahi should be eligible: https://www.sovereigntechfund.de/programs/bug-resilience/criteria#selection-criteria

I can't talk to them for various reasons (and it isn't because my downstream fuzzing infrastructure was disturbed by a bunch of changes in the systemd testing suite :-)).

**Comment by pemensik** [comment]
I admit bind9 has swiched from original xml manual pages to rst format generated by Sphinx some releases ago. And yes, it is easier to maintain and quite nice format to keep documentation in. Output is also nice and able to generate both manual pages, html and pdf too.

But I am not prepared for legal requirements for it, though yes, it could be useful to investigate more.

**Comment by evverx** [comment]
> legal requirements

My understanding is that at least in Europe the "A bug & fix bounty program" part can be problematic in the sense that it can probably conflict with employment contracts. Though as long as fix bounties are sent to charities or something like that it should be fine.

The "Direct contributions" part offloads all the stuff related to funds, contractors and things like that to STF (as far as I understand). I don't think there are any strings attached there either. At least in `systemd` pull request are opened and then eventually merged and the project itself just reviews them. (I'm not sure how `systemd` passed the "lack of support and resources" check though)

**Comment by nigels-com** [comment]
Would asciidoc be of interest for this?  That's the one I'm more familiar with.

**Comment by evverx** [comment]
I think it depends on whether it's available in all the places where avahi is shipped. As far as I know xmltoman isn't included in "main/base" parts of some distros so they just don't build the manpages. I'd stick with something that other projects shipped on Linux, *BSD, OpenIndiana and things like that use.

**Comment by nigels-com** [comment]
From a Yocto perspective, better to have the documentation in a modern format readily consumed by devs and AI.
(Nice!) HTML pages online, not man pages that depend on ancient (and unpopular) tooling, TBH.
That means using whatever toolchain makes sense to the maintainers, and not imposing that downstream.
Yes, change is hard work, understood.

**Comment by pemensik** [comment]
I would prefer python3-sphinx based documentation. Bind9 by ISC uses it for a long time, tools for building cool documentation are there. that is what generated https://bind9.readthedocs.io site. It uses Restructured Text format (*.rst). I would follow that pattern unless we have specific reason to avoid that. Sphinx tool can generate a lot of output formats, including manual pages. Advantage for me is I have seen that format and it is used in projects I work on.

**Comment by evverx** [comment]
@pemensik I really think it would make sense to contact STF. Another option would be to repurpose the OSS-Fuzz reward (it should be 5000-10000$ I think) for that purpose. The commits related to that reward are mine but I can try to talk to the OSS-Fuzz team to see whether they can be forwarded to contractors or whatever. Without that I think it isn't going to go any further.

Apart from the documentation itself It's just that everybody builds SBOMs these days and xmltoman isn't something that look good there so it all falls into those various STF things nicely as well.

**Comment by pemensik** [comment]
I am not a manager and I doubt I would ever become a good one. I hate managing my own finance legal things, let alone to be responsible for administration of non-trivial stuff in English. But I admit it should classify as valid. Although we should put together a release finally with known fixes we have.

I think we lack people willing to spend on some time. I am not ready to drive any student or even professional with list of todo items. I struggle with organizing myself often. I will try to ask our systemd people if they can share, what kinds of help they can provide and under what conditions.

**Comment by evverx** [comment]
> I will try to ask our systemd people if they can share, what kinds of help they can provide and under what conditions

Makes sense. In the context converting stuff to Sphinx and the related PR there Zbigniew would be the best person to ask

**Comment by pemensik** [comment]
Lukáš from systemd warned me that they had to do a lot of administrative work. I do not want to work on administration tasks to that. He said they had to find eliglible company, which would receive the funding and do the work. That is certainly something I want to avoid. I like coding, not administrative stuff. There is not only little needed.

I won't stop anyone going that way, but it is not something I want to do. Sorry!

**Comment by evverx** [comment]
> they had to do a lot of administrative work

That's kind of expected when hundreds of thousands of euros are allocated. I don't think the avahi scenario would be similar.

> it is not something I want to do

Understood. Let's close it as not planned then.

**Comment by pemensik** [comment]
Hmm, I think this is not a top priority, but definitely desired thing. I think even without a help from external contributors I want to attempt to do this, but not for 0.9 or 0.9.1. Lets keep it here open.

**Comment by evverx** [comment]
> I want to attempt to do this, but not for 0.9 or 0.9.1

It's all well and good but there are PRs like https://github.com/avahi/avahi/pull/828 touching that mess so maybe it would make sense to speed it up a bit. Or please show up upstream and review changes touching that mess (until it's sorted out).
